File: javascript/chatgpt/dijkstra.js
class GraphError extends Error {
  constructor(message) {
    super(message);
    this.name = "GraphError";
  }
}

/**
 * Validates adjacency matrix
 */
function validateMatrix(matrix) {
  if (!Array.isArray(matrix) || matrix.length === 0) {
    throw new GraphError("Matrix must be a non-empty 2D array.");
  }

  const n = matrix.length;

  for (let i = 0; i < n; i++) {
    if (!Array.isArray(matrix[i]) || matrix[i].length !== n) {
      throw new GraphError("Matrix must be square (NxN).");
    }

    for (let j = 0; j < n; j++) {
      const val = matrix[i][j];

      if (typeof val !== "number") {
        throw new GraphError("Matrix values must be numbers.");
      }

      if (val < 0) {
        throw new GraphError("Dijkstra does not support negative weights.");
      }
    }
  }
}

/**
 * Dijkstra algorithm using adjacency matrix
 * @param {number[][]} matrix
 * @param {number} start
 * @param {number} end
 */
function dijkstra(matrix, start, end) {
  validateMatrix(matrix);

  const n = matrix.length;

  if (
    !Number.isInteger(start) ||
    !Number.isInteger(end) ||
    start < 0 ||
    end < 0 ||
    start >= n ||
    end >= n
  ) {
    throw new GraphError("Invalid start or end node index.");
  }

  const distances = new Array(n).fill(Infinity);
  const visited = new Array(n).fill(false);
  const previous = new Array(n).fill(null);

  distances[start] = 0;

  for (let i = 0; i < n; i++) {
    let minDist = Infinity;
    let u = -1;

    for (let j = 0; j < n; j++) {
      if (!visited[j] && distances[j] < minDist) {
        minDist = distances[j];
        u = j;
      }
    }

    if (u === -1) break; // unreachable nodes remain

    visited[u] = true;

    for (let v = 0; v < n; v++) {
      const weight = matrix[u][v];

      if (weight > 0 && !visited[v]) {
        const alt = distances[u] + weight;

        if (alt < distances[v]) {
          distances[v] = alt;
          previous[v] = u;
        }
      }
    }
  }

  if (distances[end] === Infinity) {
    return {
      path: null,
      distance: Infinity,
      message: "No path exists between the given nodes."
    };
  }

  // reconstruct path
  const path = [];
  let current = end;

  while (current !== null) {
    path.unshift(current);
    current = previous[current];
  }

  return {
    path,
    distance: distances[end]
  };
}


const assert = require("assert");

// Test 1: simple graph
(function testSimpleGraph() {
  const graph = [
    [0, 2, 0, 1],
    [0, 0, 3, 0],
    [0, 0, 0, 4],
    [0, 1, 0, 0]
  ];

  const result = dijkstra(graph, 0, 2);

  assert.deepStrictEqual(result.path, [0, 1, 2]);
  assert.strictEqual(result.distance, 5);
})();

// Test 2: no path
(function testNoPath() {
  const graph = [
    [0, 1, 0],
    [0, 0, 0],
    [0, 0, 0]
  ];

  const result = dijkstra(graph, 0, 2);

  assert.strictEqual(result.path, null);
  assert.strictEqual(result.distance, Infinity);
})();

// Test 3: larger graph
(function testLargerGraph() {
  const graph = [
    [0, 4, 0, 0, 0, 0],
    [0, 0, 8, 0, 0, 0],
    [0, 0, 0, 7, 0, 4],
    [0, 0, 0, 0, 9, 14],
    [0, 0, 0, 0, 0, 10],
    [0, 0, 0, 0, 0, 0]
  ];

  const result = dijkstra(graph, 0, 5);

  assert.deepStrictEqual(result.path, [0, 1, 2, 5]);
  assert.strictEqual(result.distance, 16);
})();

// Test 4: invalid matrix
(function testInvalidMatrix() {
  try {
    dijkstra([[1, 2], [3]], 0, 1);
    assert.fail("Should throw error");
  } catch (e) {
    assert.ok(e instanceof GraphError);
  }
})();

// Test 5: negative weights
(function testNegativeWeights() {
  try {
    dijkstra([[0, -1], [0, 0]], 0, 1);
    assert.fail("Should throw error");
  } catch (e) {
    assert.ok(e instanceof GraphError);
  }
})();

console.log("All tests passed!");

File: cpp/chatgpt/dijkstra.cpp
#include <iostream>
#include <vector>
#include <limits>
#include <stdexcept>
#include <sstream>
#include <cassert>

using namespace std;

const int INF = numeric_limits<int>::max();

class GraphError : public runtime_error {
public:
    GraphError(const string& msg) : runtime_error(msg) {}
};

// ---------------------- VALIDATION ----------------------
void validateMatrix(const vector<vector<int>>& matrix) {
    if (matrix.empty()) {
        throw GraphError("Matrix must not be empty.");
    }

    size_t n = matrix.size();

    for (const auto& row : matrix) {
        if (row.size() != n) {
            throw GraphError("Matrix must be NxN.");
        }

        for (int val : row) {
            if (val < 0) {
                throw GraphError("Negative weights are not allowed.");
            }
        }
    }
}

// ---------------------- DIJKSTRA ----------------------
struct Result {
    vector<int> path;
    int distance;
    bool reachable;
};

Result dijkstra(const vector<vector<int>>& matrix, int start, int end) {
    validateMatrix(matrix);

    int n = matrix.size();

    if (start < 0 || end < 0 || start >= n || end >= n) {
        throw GraphError("Invalid start or end index.");
    }

    vector<int> dist(n, INF);
    vector<bool> visited(n, false);
    vector<int> prev(n, -1);

    dist[start] = 0;

    for (int i = 0; i < n; i++) {
        int u = -1;
        int minDist = INF;

        for (int j = 0; j < n; j++) {
            if (!visited[j] && dist[j] < minDist) {
                minDist = dist[j];
                u = j;
            }
        }

        if (u == -1) break;

        visited[u] = true;

        for (int v = 0; v < n; v++) {
            if (matrix[u][v] > 0 && !visited[v]) {
                int alt = dist[u] + matrix[u][v];

                if (alt < dist[v]) {
                    dist[v] = alt;
                    prev[v] = u;
                }
            }
        }
    }

    if (dist[end] == INF) {
        return {{}, INF, false};
    }

    vector<int> path;
    for (int at = end; at != -1; at = prev[at]) {
        path.insert(path.begin(), at);
    }

    return {path, dist[end], true};
}

// ---------------------- OUTPUT ----------------------
void printResult(const Result& res) {
    if (!res.reachable) {
        cout << "No path exists.\n";
        return;
    }

    cout << "Path: ";
    for (int node : res.path) {
        cout << node << " ";
    }
    cout << "\nDistance: " << res.distance << endl;
}

// ---------------------- UNIT TESTS ----------------------
void runTests() {
    // Test 1: simple graph
    {
        vector<vector<int>> g = {
            {0, 2, 0, 1},
            {0, 0, 3, 0},
            {0, 0, 0, 4},
            {0, 1, 0, 0}
        };

        auto res = dijkstra(g, 0, 2);
        assert(res.reachable);
        assert(res.distance == 5);
        assert((res.path == vector<int>{0,1,2}));
    }

    // Test 2: no path
    {
        vector<vector<int>> g = {
            {0, 1, 0},
            {0, 0, 0},
            {0, 0, 0}
        };

        auto res = dijkstra(g, 0, 2);
        assert(!res.reachable);
    }

    // Test 3: larger graph
    {
        vector<vector<int>> g = {
            {0, 4, 0, 0, 0, 0},
            {0, 0, 8, 0, 0, 0},
            {0, 0, 0, 7, 0, 4},
            {0, 0, 0, 0, 9, 14},
            {0, 0, 0, 0, 0, 10},
            {0, 0, 0, 0, 0, 0}
        };

        auto res = dijkstra(g, 0, 5);
        assert(res.reachable);
        assert(res.distance == 16);
        assert((res.path == vector<int>{0,1,2,5}));
    }

    // Test 4: invalid matrix
    {
        try {
            vector<vector<int>> g = {{1,2},{3}};
            dijkstra(g, 0, 1);
            assert(false);
        } catch (const GraphError&) {}
    }

    // Test 5: negative weights
    {
        try {
            vector<vector<int>> g = {{0,-1},{0,0}};
            dijkstra(g, 0, 1);
            assert(false);
        } catch (const GraphError&) {}
    }

    cout << "All tests passed!\n";
}

// ---------------------- MAIN ----------------------
int main() {
    try {
        runTests();

        int n;
        cout << "Enter number of nodes: ";
        cin >> n;

        if (n <= 0) throw GraphError("Invalid graph size.");

        vector<vector<int>> matrix(n, vector<int>(n));

        cout << "Enter adjacency matrix:\n";
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                cin >> matrix[i][j];
            }
        }

        int start, end;
        cout << "Start node: ";
        cin >> start;
        cout << "End node: ";
        cin >> end;

        auto result = dijkstra(matrix, start, end);
        printResult(result);

    } catch (const GraphError& e) {
        cerr << "Graph error: " << e.what() << endl;
    } catch (const exception& e) {
        cerr << "Unexpected error: " << e.what() << endl;
    }

    return 0;
}
